Transform netfile data to a database schema useful for APIs.

bcipolli opened the pull request:
This PR tries to take netfile data and shove it into the schema specified in https://github.com/caciviclab/disclosure-backend/pull/103 . Trying to do this has made me change that schema time and time again, which is great!

Currently done:
- Validate the `finance` schema--it seems to be working well.
- Do a reasonable job creating different models for different types of benefactors (those contributing and spending on behalf of a ballot measure choice/candidate), such as individuals, companies, committees
- Do a reasonable job filling in data for beneficiaries (committees).
- Do a reasonable job saving contribution data from Form 460A.

To do before merge:
- [x] Scrape locality data for beneficiary, ballot models.
- [x] Scrape candidate/ballot measure name for beneficiary, ballot measure choice models.
- [x] Figure out what `filingId` is.
- [x] Add `help_text` for each model field
- [x] Fix default (`None` vs. empty) for each model field.

To do after merge (will open issues after merge):
- Find some info indicating support/oppose
- Find some info about reporting period start/end from the data.
- Get form 496 data working. (better to get this in with limited data, so that we can get APIs up & running)
- Get form 497 data working (harder, because timing matters)
- Alert admins when new ballot info has been added. It could be error-prone or have missing data, and needs to be manually checked.
- Alert admins when csv rows have failed to be saved. Each failed row should be stored in the database and checked manually.
- Fix up data issues (e.g. zip codes can be 5 digit or 10 digit (5-4 format))

I have run this on a subset of the data (ANA, 2015), and am currently running it on all the data (68 agencies, 2014-2015). 

@mrmaksimize let's work on this branch together, yes?

bcipolli commented:
Note that the only commit that is _actually_ on this branch, is the final one. All the rest are commits from other branches, that will disappear when this branch gets rebased.

Commit:
ENH: transform netfile data to our database schema.

Commit:
RF: split function pieces into small sub-functions.

Commit:
ENH: set beneficiary and ballot_measure locality properly.

Commit:
ENH: determine ballot date from transaction date.

Commit:
fix: xform

Commit:
ENH: parse OfficeElection/Referendum, and Candidate/Referendum responses

bcipolli commented:
OK, this is working pretty well. After populating with San Diego 2015 Form 460 Schedule A data, some sample queries:

```
mysql> select * from ballot_ballotitem;
+----+--------------+---------------------------------------------------------------------+--------+-----------+
| id | contest_type | name                                                                | number | ballot_id |
+----+--------------+---------------------------------------------------------------------+--------+-----------+
|  1 | O            | Election of City Council in San Diego, City of                      |        |         1 |
|  2 | R            | Unknown                                                             |        |         1 |
|  3 | O            | Election of Council in San Diego, City of                           |        |         1 |
|  4 | O            | Election of Tomorrow's Economy supporting the in San Diego, City of |        |         1 |
|  5 | O            | Election of  in San Diego, City of                                  |        |         1 |
|  6 | O            | Election of Assembly in San Diego, City of                          |        |         1 |
|  7 | O            | Election of Mayor in San Diego, City of                             |        |         1 |
|  8 | O            | Election of City Council District in San Diego, City of             |        |         1 |
|  9 | O            | Election of City Attorney in San Diego, City of                     |        |         1 |
| 10 | O            | Election of For District in San Diego, City of                      |        |         1 |
| 11 | O            | Election of San Diego City Council District in San Diego, City of   |        |         1 |
| 12 | O            | Election of For City Council in San Diego, City of                  |        |         1 |
| 13 | O            | Election of San Diego City Attorney in San Diego, City of           |        |         1 |
+----+--------------+---------------------------------------------------------------------+--------+-----------+
```

```
mysql> select * from finance_committee limit 5;
+----+-----------+-------------+--------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+-------------+----------+------+
| id | photo_url | website_url | facebook_url | twitter_url | name                                                                                                                                                             | address_id | locality_id | filer_id | type |
+----+-----------+-------------+--------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+-------------+----------+------+
|  1 |           |             |              |             | Mitz Lee for City Council 2014                                                                                                                                   |       NULL |           4 | 152867   | PF   |
|  2 |           |             |              |             | San Diego Small Business Coalition sponsored by the San Diego Regional Chamber of Commerce                                                                       |       NULL |           4 | 153357   | PF   |
|  3 |           |             |              |             | Neighborhoods for Clean Elections                                                                                                                                |       NULL |           4 | 10101    | PF   |
|  4 |           |             |              |             | Myrtle Cole Legal Defense Fund                                                                                                                                   |       NULL |           4 | 153339   | PF   |
|  5 |           |             |              |             | Protect San Diego's Neighborhoods, a coalition of neighborhood residents, community planners, taxpayers, and small businesses.  Major funding by Donahue Shriber |       NULL |           4 | 153573   | PF   |
+----+-----------+-------------+--------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+-------------+----------+------+
5 rows in set (0.00 sec)
```

```
mysql> select * from finance_independentmoney limit 5;
+----+----------+---------+-------------------+---------+---------------------+---------------+----------------+-------------------------+-----------+--------+----------------------------------+
| id | amount   | support | benefactor_zip_id | form_id | reporting_period_id | benefactor_id | beneficiary_id | ballot_item_response_id | filing_id | source | source_xact_id                   |
+----+----------+---------+-------------------+---------+---------------------+---------------+----------------+-------------------------+-----------+--------+----------------------------------+
|  1 |     3000 |       1 |                 3 |       1 |                   1 |             1 |              1 |                       1 | 154002316 | NF     | 737b897fa80842229919a4340105967b |
|  2 | 12032.56 |       1 |                 3 |       1 |                   2 |             2 |              1 |                       1 | 154002316 | NF     | 7587504c4a484a6caeb7a4340105967b |
|  3 |     5000 |       1 |                 6 |       1 |                   3 |             3 |              2 |                       2 | 154571178 | NF     | f6d4d9bff264427f8980a45d00af1703 |
|  4 |     4950 |       1 |                 7 |       1 |                   4 |             4 |              2 |                       2 | 154571178 | NF     | 1d693ac500ed472f8e19a45d00af1703 |
|  5 |      100 |       1 |                 8 |       1 |                   5 |             5 |              3 |                       2 | 155017741 | NF     | 11e7a0c9d7d943b2875aa48300a4de02 |
+----+----------+---------+-------------------+---------+---------------------+---------------+----------------+-------------------------+-----------+--------+----------------------------------+
```

bcipolli commented:
The only major issue I have with this code (other than it's incorrect and incomplete, but that'll take weeks to fix) is:
- There is no `ReferendumResponse` object. I manually create a `BallotItemResponse` with a "YES" answer, when I guess that something's related to a Referendum.

Otherwise, I think this is pretty rockin'! It's good enough that we should be able to enable live JSON APIs for front-end developers.

bcipolli commented:
Here's what the `BallotItem` data look like when I add in Oakland 2015. Not perfect, but damn if it doesn't make some sense! :) Even got different `ballot_id`s...

![image](https://cloud.githubusercontent.com/assets/4072455/12223929/2664c4dc-b799-11e5-9b68-5b1a74e1ff1f.png)

adborden commented:
Getting an error with the migration..

```
$ mysql "CREATE DATABASE opendisclosure;"
$ python manage.py flush --database calaccess_raw
$ python manage.py migrate
```

```
$ python manage.py migrate
Operations to perform:
  Synchronize unmigrated apps: disclosure, corsheaders, finance, zipcode_metro_raw, locality, staticfiles, office_election, election_day, messages, referendum, calaccess_raw, rest_framework_swagger, rest_framework
  Apply all migrations: admin, contenttypes, auth, sessions
Synchronizing apps without migrations:
  Creating tables...
    Running deferred SQL...
  Installing custom SQL...
Running migrations:
  Rendering model states...Traceback (most recent call last):
  File "manage.py", line 11, in <module>
    execute_from_command_line(sys.argv)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/core/management/__init__.py", line 338, in execute_from_command_line
    utility.execute()
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/core/management/__init__.py", line 330, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/core/management/base.py", line 390, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/core/management/base.py", line 441, in execute
    output = self.handle(*args, **options)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 221, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/db/migrations/executor.py", line 100, in migrate
    state.apps  # Render all real_apps -- performance critical
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/utils/functional.py", line 60, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/db/migrations/state.py", line 154, in apps
    return StateApps(self.real_apps, self.models)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/db/migrations/state.py", line 220, in __init__
    self.render_multiple(list(models.values()) + self.real_models)
  File "/home/adborden/.virtualenvs/disclosure-backend/lib/python2.7/site-packages/django/db/migrations/state.py", line 258, in render_multiple
    "for more" % (new_unrendered_models, get_docs_version())
django.db.migrations.state.InvalidBasesError: Cannot resolve bases for [<ModelState: 'office_election.OfficeElection'>, <ModelState: 'office_election.Candidate'>, <ModelState: 'referendum.Referendum'>]
This can happen if you are inheriting models from an app with migrations (e.g. contrib.auth)
 in an app with no migrations; see https://docs.djangoproject.com/en/1.8/topics/migrations/#dependencies for more
```

bcipolli commented:
No, looks like I fouled something up in the `refactor` branch with the last squash I did. Lemme fix.

bcipolli commented:
@adborden Actually no, it's just stale `.pyc` files (at least on my machine). Try running:
`find . -name \*.pyc -exec rm \{\} \;`

adborden commented:
Great, looking better!

Commit:
TST: add xformnetfiledata test.

bcipolli commented:
@adborden I just removed the multi-db stuff from our side, made an explicit dependency on a version of my `calaccess_raw` module, and added a couple of tests. I had to rebase to do it, so you'll have to `stash/checkout/stash pop` to get the changes. On the other hand, up to 83.5% test coverage! :)

Commit:
ENH: save each form row entry data in a transaction.

Commit:
fix: update for schema changes.

Commit:
BF: for now, assume one reporting period per year.

Commit:
STY: remove debug-only code.

bcipolli commented:
@adborden I updated this with a cleaned schema from #103, cleaned code, better "faking" of reporting period info, etc.

Commit:
RF: make address a mixin.

bcipolli commented:
OK, I marked this PR as ready to go. With merging the other branches, this one becomes very clean.

The logic may not be 100% correct, but it populates the DB with enough data to make adding data feeds relatively simple and unblock the UI work.

Because of this, I recommend a relatively quick review, so we can get the front-end side moving along for @jnmarcus & company.

tdooner commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+                raise
+                if self.verbosity:
+                    print("%d: \n\n%s\n\n" % (len(error_rows) + 1, e))
+                    time.sleep(0.5)
```
huh?

bcipolli replied:
Haha debug code; it flashes by fast if you don't do this. This can all go away.

tdooner commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+        # Report errors  TODO: push to the database.
+        if len(error_rows) > 0:
+            if self.verbosity:
+                print("Encountered %d errors; debug!" % len(error_rows))
```
I think we should print this even if we aren't in verbose mode.

bcipolli replied:
:+1:

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+                len(self.data), self.combined_csv_path))
+
+        # Check for any potentially missing data.
+        form_types = self.data['form_Type'].unique()
```
Where does the casing come from? can we just normalize (lowercase) the column names?

bcipolli replied:
It's native to the csv. I'd prefer to mirror the names in the CSV, rather than manipulate them. I worry that manipulating the column names to be lowercase could confuse people familiar with the data. And it'd be a pain in the butt :)

If you feel this is an important point, I can convert the column labels to lowercase, just lemme know.

tdooner commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+
+                # of=official, pf=primarily, ic=independent
+                benefactor.name = row['tran_NamL'].strip()
+                benefactor.type = benefactor.type or 'PF'  # TODO: fix
```
fix what?

tdooner commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+                '^(?P<name>.*?)\s+for\s+(?P<office>.*?)(?P<year>[0-9]+).*$',
+                # Scott Sanborn City Council 2016
+                '^(?P<name>.*? .*?)\s+(?P<office>.*?)(?P<year>[0-9]+).*$',
+            )
```
Does this actually get good results?

adborden replied:
This is a placeholder? Makes sense. I'd like to see some TODO comments so folks looking at this understand what pieces need to be built.

mikeubell replied:
There was one candidate for Mayor in Oakland who did not have her full name in the committee name.  I don't think we can count on all committee names to follow a known format.  On the other hand most will so maybe we just need to keep an eye out for those that do not.

> On Jan 13, 2016, at 6:52 PM, Tom Dooner notifications@github.com wrote:
> 
> In finance/management/commands/xformnetfilerawdata.py:
> 
> > -            beneficiary.locality = locality
> > -            beneficiary.name = row['filerName'].strip()
> > -            beneficiary.type = 'PF'  # ok
> > -            # beneficiary.address = '?'  # TODO: fix
> > -            beneficiary.save()
> > -            return beneficiary
> > -        beneficiary = parse_beneficiary(row)
> >   +
> > -        def parse_candidate_and_office(row):
> > -            import re
> > -            res = (
> > -                # David Alvarez for Mayor 2014
> > -                '^(?P<name>._?)\s+for\s+(?P<office>._?)(?P<year>[0-9]+).*$',
> > -                # Scott Sanborn City Council 2016
> > -                '^(?P<name>._? ._?)\s+(?P<office>._?)(?P<year>[0-9]+)._$',
> > -            )
> >   Does this actually get good results?
> 
> —
> Reply to this email directly or view it on GitHub.

bcipolli replied:
@mikeubell Thanks for the note--yes, I've seen a lot of things like that as well. @tdooner This works reasonably well, definitely not perfectly. The goal here was to get reasonable data into the database--not correct, but enough that we can build out the API endpoints and have live data for the front-end folks to work with. Then, we can refine this type of back-end code along the way in parallel.

tdooner commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+        # Split data by type, add column defs.
+        self.f496_data = self.data[self.data['form_Type'] == 'F496P3']
+        if self.verbosity:
+            print("NYI: load %d rows of form 496 data." % len(self.f496_data))
```
So, uh, that's a lot of code. Since testing is cool, would you mind adding one or two tests of code in this module? That is, just to get the framework in place, and then we can fill in more tests later. At least, it always seems to me like writing the first test is the hardest, so I want to saddle you with that burden if possible :grinning:

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+        money.save()
+
+        if self.verbosity:
+            print(str(money))
```
does django management provide a logging facility?

tdooner commented:
This looks like a good approach in general, and I think breaking out utility methods like all the `parse_` methods will make it pretty easy to audit and test the logic we apply to import the data.

I didn't see any logic to handle amendments, I suppose that can come later if we need it.

:+1:

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+        """
+        filer_id = row['filerId']
+        if '-' in filer_id:
+            filer_id = '-'.join(filer_id.split('-')[1:])
```
What's this for? drop the prefix?

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+        if '-' in filer_id:
+            filer_id = '-'.join(filer_id.split('-')[1:])
+        else:
+            filer_id = filer_id
```
Is this case necessary?

bcipolli replied:
lol remnants of my search-replace. Removed.

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+
+    @staticmethod
+    def clean_filer_id(filer_id):
+        """ Utility function to scrub committee IDs."""
```
s/committee/filer

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+
+    def load_f497_data(self):
+        """ Loads data from Form 460 Schedule A:
+        contributions to primarily formed committees."""
```
Loads F497 from F460A, or Loads F497 data?

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+        # Split data by type, add column defs.
+        self.f497_data = self.data[self.data['form_Type'] == 'F497P2']
+        if self.verbosity:
+            print("NYI: load %d rows of form 497 data." % len(self.f497_data))
```
NYI? not yet implemented?

adborden commented on finance/management/commands/xformnetfilerawdata.py:
```diff
@@ -0,0 +1,365 @@
+            return isinstance(val, Number) and np.isnan(val)
+
+        def isnone(val):
+            return val is None or val == 'None'
```
`isnan` and `isnone` are probably good to move to the bottom with the other utility functions?

adborden commented:
```
$ python manage.py  xformnetfilerawdata --agencies COAK --years 2015
```

Gives me [this error](https://gist.github.com/adborden/8c0b3cbfc3c9543df256).

bcipolli commented:
@adborden Probably a database error? Try drop/recreate database and run `migrate`?

adborden commented:
Here's my steps:

```
$ pip install -r requirements.txt
$ find . -name \*.pyc -exec rm \{\} \;
$ python manage.py flush
$ python manage.py  xformnetfilerawdata --agencies COAK --years 2015
```

adborden commented:
Ah, my bad. `flush` does not clear out the tables like I had though. Is there a django command that WILL get you to a "fresh install" state?

Commit:
RF: move parse functions off of Command object.

Commit:
RF: make tests directory, stub parse_ tests.

Commit:
fix

Commit:
RF: Add ReferendumResponse, move props from BallotItemResponse

Commit:
TST: add tests for Candidate election sleight-of-hand.

Commit:
RF: BallotItemResponse => BallotItemSelection

Commit:
fix: schema

Commit:
fix: tests

Commit:
fix: schema

Commit:
fix: officeelection

Commit:
fix: tests

Commit:
TST: add tests for isnan, isnone.

bcipolli commented:
@tdooner @adborden I think this is ready to go. I broke out all the utility functions to be easily testable, tested two of the functions, and made it clear where more tests need to go. 

I think more testing can come later, as all this logic is going to change as we improve the parsing (that's what I've seen as I made API endpoints live).

bcipolli commented:
Keeping this in PR is doing me more harm (trying to maintain multiple branches) than good, so will merge. Please complain later if you disagree, and I'll be more patient! :)
